core/cache/modelsummarywatcher_test.go

```diff
@@ -0,0 +1,578 @@
+		},
+	}, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:      "model-3-uuid",
+			Namespace: "mary",
+			Name:      "model-3",
+			Admins:    []string{"mary"},
+			Status:    cache.StatusGreen,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestRemoveModelRemovesModel(c *gc.C) {
+	watcher := s.controller.WatchModelsAsUser("bob")
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the initial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.RemoveModel{
+		ModelUUID: "model-2-uuid",
+	}, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:    "model-2-uuid",
+			Removed: true,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestAddingMachineIsChange(c *gc.C) {
+	watcher := s.controller.WatchAllModels()
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the intial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.MachineChange{
+		ModelUUID: "model-2-uuid",
+		Id:        "0",
+		Life:      life.Alive,
+	}, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:         "model-2-uuid",
+			Namespace:    "bob",
+			Name:         "model-2",
+			Admins:       []string{"bob"},
+			Status:       cache.StatusGreen,
+			MachineCount: 1,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestRemovingMachineIsChange(c *gc.C) {
+	watcher := s.controller.WatchAllModels()
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the intial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.RemoveMachine{
+		ModelUUID: "model-1-uuid",
+		Id:        "0",
+	}, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:      "model-1-uuid",
+			Namespace: "test-admin",
+			Name:      "model-1",
+			Admins:    []string{"test-admin"},
+			Status:    cache.StatusGreen,
+			// We didn't actually remove the application, or unit yet.
+			ApplicationCount: 1,
+			UnitCount:        1,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestAddingApplicationIsChange(c *gc.C) {
+	watcher := s.controller.WatchAllModels()
+	defer workertest.CleanKill(c, watcher)
+
+	changes := watcher.Changes()
+	// discard the intial event
+	_ = s.next(c, changes)
+
+	s.ProcessChange(c, cache.ApplicationChange{
+		ModelUUID: "model-2-uuid",
+		Name:      "foo",
+		Life:      life.Alive,
+	}, s.events)
+
+	update := s.next(c, changes)
+	c.Assert(update, jc.DeepEquals, []cache.ModelSummary{
+		{
+			UUID:             "model-2-uuid",
+			Namespace:        "bob",
+			Name:             "model-2",
+			Admins:           []string{"bob"},
+			Status:           cache.StatusGreen,
+			ApplicationCount: 1,
+		},
+	})
+}
+
+func (s *modelSummaryWatcherSuite) TestRemovingApplicationIsChange(c *gc.C) {
```

Comment:
what about changing an applications Charm URL?

Reply:
Changing an application's charm URL doesn't actually change the model itself... yet.

There has been an ask for a dateLastModified to the model itself, which would change on any write, but we dont have that yet.

Reply:
By not changing the model, I mean not changing the actual model summary.